### PR TITLE
Initial Development (v1.0.0)

### PR DESCRIPTION
**Pull request by localjo**
use solidus-client and wp nonce to render preview into default template

### CODE REVIEW

**Comment by pushred** [comment]
I’m not sure about the scope of this module.. What’s really throwing me off is the example template. If you’re going to provide that, why not just call this the beginning of a module for post content, rather than limiting it to previews? There’s nothing special about the post template just for the purpose of that. If we added a post module later this one would just be extraneous.

Aside from the template it’s really just a utility module for making authenticated clientside requests to WordPress and overwriting an element. Expanding on the changes to Solidus maybe this is better off as a `solidus-preview` module that we can expand later to simplify accessing unpublished content in other services. Universe could be next. Going that route I think the example template should be dropped and the option should just direct users to point to the corresponding template that’s used to render the content for public visitors.

File: src/post-preview.js [hunk 1]
```diff
@@ -14,15 +17,35 @@ function PostPreview (options) {
...
+          url: options.resource,
+          with_credentials: true,
+          query: {'_wp_json_nonce': params._wp_json_nonce}
+        }
+      },
+      preprocessor: preprocessor,
+      template: instance.template
+    };
+
+    previewClient.render(view).end(function(err, html) {
```
Comment (pushred) [inline src/post-preview.js]: Should throw the error if there is one here

File: src/post-preview.js [hunk 2]
```diff
@@ -14,15 +17,35 @@ function PostPreview (options) {
...
+      if (html) {
+        instance.element.outerHTML = html;
         instance.element = document.getElementById(options.id);
```
Comment (pushred) [inline src/post-preview.js]: What’s this do?

File: README.md [hunk 3]
```diff
@@ -32,9 +32,11 @@ The module can be configured with several optional parameters passed to it's con
 ### `element`
 **(Required)** A DOM element, jQuery element, or selector string to refer to the placeholder element.
```
Comment (pushred) [inline README.md]: Should tweak the docs to stress this is a bit destructive, replacing the specified element.

**Comment by pushred** [comment]
So backing up before I look at code, can you respond to my original comment here in the context of where you're at now?

**Comment by localjo** [comment]
Re: that comment, I agree with the direction you're going. I removed the default template, because it doesn't make sense with that approach. I'm not sure if I have a clear idea of what you mean by a "post module" that we'd add later. In regards to naming, `solidus-preview` and `post-preview` seem really similar to me. Could be `solidus-post-preview` too, but I guess what you're getting at is that it shouldn't be specific just to "post" content, but all types of content?

**Comment by pushred** [comment]
Yup, but what I'm really talking about here is what's a block module and what isn't. To me they are essentially modern widgets; they provide a discrete capability to end users and offer a quick way to get something on a page while offering plenty of ways to customize as needed. The reason they're modules is so that you can compose a site's features (or roll them up for groups of similar sites) without going the monolithic route.

This stuff also provides a capability, but it's technical in nature. It's rolling up some configuration and creating some abstracted methods that layer functionality on top of an underlying foundation (solidus). I think generally you could call this adapter code but I'm also looking at it as a plugin, much in how gulp works, or some other new favorite tools of mine, [PostCSS](https://github.com/postcss/postcss) and [markdown-it](https://github.com/markdown-it/markdown-it), or even [express](http://expressjs.com/guide/using-middleware.html) and [Hapi](http://hapijs.com/plugins). I'd really like to take solidus in that pluggable direction, breaking up our solution so it's more flexible and doesn't come with quite so many dependencies. solidus-client is a good first step in that direction. This could be another.

Both [Contentful](https://www.contentful.com/developers/docs/references/content-preview-api/#/reference) and [Prismic.io](https://blog.prismic.io/VGTPsyQAACQAqzsZ/preview-content-changes-in-your-website) have had to put together some similar preview functionality that we would use directly if we ever used those services. But most APIs won't offer us those luxuries, so this won't be the last such adapter we need to write.

If the complexity ultimately gets too much for one module you could go a route like [OmniAuth strategies](https://github.com/intridea/omniauth/wiki/List-of-Strategies) 
but I don't think it will be like that for a very long time. The modules would get out of hand, which [needlessly adding more complexity](http://hueniverse.com/2014/05/30/the-fallacy-of-tiny-modules/) for our users without [giving them any benefit](https://medium.com/@Rich_Harris/small-modules-it-s-not-quite-that-simple-3ca532d65de4). A better example would be from the node corollary, [authom][https://github.com/jed/authom]. It's a super fine line that we need to figure out better, but that will require much more thinking on another day.

Since this is so light, something you could consider is just including this in the solidus package itself, which already includes solidus-client. Thoughts?

cc @joanniclaborde

**Comment by localjo** [comment]
I think in order to respond to the ideas you're bringing up, I need a clearer idea of the goals we're trying to accomplish and the scope of this. I've been building this based on the goals laid out in the Trello card (https://trello.com/c/zElId0eA/1074-wordpress-post-previews). Essentially the problem I'm solving at this point is "Rendering unpublished content into a Solidus site template so that content editors can immediately see their changes." Packaging this as a module allows us to use the solution across any Solidus site without having to repeat ourselves. This module solves that problem pretty well, as-is, but if we're trying to solve a broader problem, let's define that problem more clearly so I can focus on that.

**Comment by joanniclaborde** [comment]
Isn't this just another flavour of [universe-js](https://github.com/SparkartGroupInc/universe-js)? A module that renders templates with resources that require WordPress client-side authentication (instead of Universe).

**Comment by pushred** [comment]
Well the goal really is to just find a home for this code — the blocks org isn't it. I'm sorry if my long-winded answer masked that, I was just trying to expand my thoughts on this as a `solidus-preview` module. I'm not talking about any functional changes, but rather just staking a place to put this sort of code going forward.

Yes it's like universe.js in that it's a wrapper/adapter for an API. The difference is that this covers a couple GET endpoints fetched for the specific purpose of preview rather than being a general purpose WP-API client.

It could very well be a module on it's own, I just brought up including it in Solidus proper out of concern for tiny modules. If it became more substantial I'd motion for `solidus-preview` again.

**Comment by joanniclaborde** [comment]
I'm not sure about including this in solidus or even solidus-client, we don't really want to release new versions of those whenever API-specific code changes (WordPress in this case). Perhaps a new module like solidus-tools or solidus-client-helpers (since it's front-end code) could contain all such functionalities.

If we do go with a new module containing multiple solutions like post-preview (which I would rename into wordpress-post-preview), just make sure all solutions are not requiring each other in any way. We don't want browserify to package all the module...

**Comment by localjo** [comment]
To me, it would make sense to move this module over to the solidusjs org, and include it as a dependency in solidus-client, so that solidus-client has preview functionality out of the box. I'm leaning toward one preview module for all of the APIs we want to preview, and an implementation that would look something like this;

```
var solidus_client = require('solidus-client');
var post_template = require('../../../views/modules/story.hbs');

module.exports = function(){ 
  var solidus_client = new SolidusClient();
  solidus_client.preview({
    service: 'wordpress',
    domain: 'http://wordpress.jasonaldean.com',
    element: '.story.module',
    post_id: solidus.parameters.id,
    template: post_template,
    preprocessor: preprocessor,
    template_options: {}
  });
  return solidus_client;
};
```

That's just a rough idea to show the direction I'm thinking. Specifics might change a little.

**Comment by localjo** [comment]
So, talking with @joanniclaborde it seems like a better approach may be a different function for each API;

```
SolidusClient.previewWordPress({
    domain: 'http://wordpress.jasonaldean.com',
    element: '.story.module',
    post_id: solidus.parameters.id,
    template: post_template,
    preprocessor: preprocessor,
    template_options: {}
  })
```

And the implementation of each of those separate API functions could be in a `solidus-preview` utility module that's a dependency of `solidus-client`.

**Comment by joanniclaborde** [comment]
Maybe we should drop the `element` argument, and just return the rendered HTML, like SolidusClient does. If we want to support that feature, we should add it to SolidusClient itself.

**Comment by localjo** [comment]
One of our goals is to make it as easy as possible for developers to implement preview functionality for content editors. I don't think there's any variation in what's done with the resulting HTML here; it's all rendered into the post content element. I think it's easier to abstract that so that developers don't have to think about the implementation every time.

**Comment by joanniclaborde** [comment]
Definitely, that's why it could be added to the original SolidusClient render method.

**Comment by localjo** [comment]
Moving this to https://github.com/solidusjs/solidus-preview/pull/1
And https://github.com/solidusjs/solidus-client/pull/10

**Pull request closed without merging**